velox/functions/prestosql/StringFunctions.h

```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
 };
 
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
+/// to get word stem
+class Stemmer {
+ private:
+  sb_stemmer* sbstemmer;
+  Stemmer(sb_stemmer* stemmer) : sbstemmer(stemmer) {}
+
+ public:
+  ~Stemmer() {
+    sb_stemmer_delete(sbstemmer);
+  }
+
+  /// Get a Stemmer from the the map stored in thread local storage
+  /// or create a new one if it doesn't exist. Return NULL if the
+  /// specified lang is not supported.
+  static Stemmer* getStemmer(const char* lang) {
+    thread_local std::map<std::string, Stemmer*> stemmers;
+    if (auto found = stemmers.find(lang); found != stemmers.end()) {
+      return found->second;
+    }
+    // Only support ASCII and UTF-8
+    if (auto sbstemmer = sb_stemmer_new(lang, "UTF_8")) {
+      auto rev = new Stemmer(sbstemmer);
+      stemmers[lang] = rev;
+      return rev;
+    }
+    return NULL;
+  }
+
+  /// Get the word stem or NULL if out of memory
+  const char* stem(const std::string& input) {
+    return (const char*)(sb_stemmer_stem(
+        sbstemmer,
+        reinterpret_cast<unsigned char const*>(input.c_str()),
+        input.length()));
+  }
+};
+
+/// word_stem function
+/// word_stem(word) -> varchar
+///     return the stem of the word in the English language
+/// word_stem(word, lang) -> varchar
+///     return the stem of the word in the specificed language
+///
+/// It uses the snowball stemmer library to calculate the stem.
+/// https://snowballstem.org
+/// It provides Java implementation which is used in Presto as well
+/// as C implementation. Therefore, both Presto and Prestimissio
+/// would have the same word stem results.
+template <typename T>
+struct WordStemFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input) {
+    return doCall<false>(result, input);
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input) {
+    return doCall<true>(result, input);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& lang) {
+    return doCall<false>(result, input, lang.data());
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& lang) {
+    return doCall<true>(result, input, lang.data());
+  }
+
+  template <bool isAscii>
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const char* lang = "en") {
+    auto stemmer = Stemmer::getStemmer(lang);
+    if (!stemmer) {
+      // language is not supported
+      throw std::invalid_argument(
```

Comment:
use VELOX_USER_FAIL

Reply:
good tip. let me also explore other macros for future use. thanks.